### ISSUE TITLE
[Performance][MEDIUM] General performance investigation for opening app from customer profile trace

### ISSUE THREAD

**Issue by hannojg** [Expensify/App#46595]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?

We had a customer meeting and they had really bad performance. We recorded multiple profile traces with them.
This issue is about the profile trace for app open.

The collection of issues:

- https://github.com/Expensify/App/issues/46590
- [Partially backend issue] https://github.com/Expensify/App/issues/46591
- https://github.com/Expensify/App/issues/46593
- https://github.com/Expensify/App/issues/46594
- https://expensify.slack.com/archives/C049HHMV9SM/p1722434442673559
 - https://expensify.slack.com/archives/C049HHMV9SM/p1722438425183629

## What is the impact of this on end-users?

Slow or completely hanging/blocked app when starting.

## List any benchmarks that show the severity of the issue

https://share.firefox.dev/4fgLVcm

## Proposed solution (if any)

None yet, I will go through the profile and see what can be optimised, what exactly caused those lags.

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)

not available yet 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

**Version Number:** v9.0.11-5
**Reproducible in staging?:** not tested
**Reproducible in production?:** yes
**Email or phone of affected tester (no customers):** customer
**Logs:** See performance file
**Notes/Photos/Videos:** See attached video
**Expensify/Expensify Issue URL:** n/a
**Issue reported by:** @hannojg
**Slack conversation:** https://expensify.slack.com/archives/C05LX9D6E07/p1721919928992729

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by hannojg** [comment]
cc @sakluger (feel free to assign me as I (or someone from my team) will work on this ticket!)

**Comment by sakluger** [comment]
Hey @hannojg, any updates on this collection of issues?

**Comment by hannojg** [comment]
Working through the issues one by one, currently working on this one:

- https://github.com/Expensify/App/issues/46591

I will also find someone from my team to help with these performance issues!

**Comment by sakluger** [comment]
Looks like we are adding some timing performance logging: https://github.com/Expensify/App/pull/46807

**Comment by sakluger** [comment]
I'm going to change this to a Weekly since it's a tracking issue related to four specific performance investigations.

**Comment by hannojg** [comment]
One important thing we just found out.
So someone from our team created in old dot ~2000 reports. When I opened new dot OpenApp took in total ~3 min to complete (and also returned over 20mb of data).

During that time the whole app seemed block, presumably due to the sequential queue waiting for OpenApp to finish.

I feel like this is something we maybe want to improve on the backend?

During that time I was seeing a constant skeleton loader (I actually thought the app would be broken).

cc @chrispader

**Comment by sakluger** [comment]
Hanno distributed the issues above to a few other people so we can work on them in parallel. Otherwise, no major updates. I'll be OOO for the next week so I will check back then.

**Comment by muttmuure** [comment]
How are these coming along @hannojg ?

**Comment by sakluger** [comment]
Hey @hannojg, do you have any updates on any of the performance issues we're tracking here?

**Comment by hannojg** [comment]
We are working on a new PR here which improves the search speed by **2456x** (344ms to 0.14ms):

- https://github.com/Expensify/App/pull/48652

I am OOO from 09.09-22.09, please assign @kirillzyusko meanwhile as he will take care of finishing the PR while I am away!

**Comment by sakluger** [comment]
Hey @kirillzyusko could you please leave a comment here so I can assign you to the issue? Thanks!

**Comment by kirillzyusko** [comment]
@sakluger yes, sure, feel free to assign me on this ticket 👋

**Comment by sakluger** [comment]
No updates. @kirillzyusko have you been working on that WIP PR, or are we waiting for Hanno to return?

**Comment by kirillzyusko** [comment]
@sakluger yeah, I've been working. I cleaned up the implementation, tested more carefully and discovered, that a new implementation produces a different search output (i. e. sometimes it doesn't find users). Currently investigating that and trying to fix!

**Comment by sakluger** [comment]
Hey @hannojg and @kirillzyusko, any updates on the WIP PR?

**Comment by hannojg** [comment]
Provided an update here:

- https://github.com/Expensify/App/issues/46591#issuecomment-2374350034

PR should become ready today!!

**Comment by sakluger** [comment]
Looks like steady progress is being made in the PR review.

**Comment by sakluger** [comment]
Looks like all reviews are complete on https://github.com/Expensify/App/pull/48652, but it's failing one test. Hopefully this will be live very soon!

**Comment by sakluger** [comment]
The PR was merged but seems to have caused a deploy blocker: https://github.com/Expensify/App/issues/51123

**Comment by muttmuure** [comment]
This went to prod!

**Comment by hannojg** [comment]
Hm, did it? I think the PR was reverted

**Comment by hannojg** [comment]
I am currently working on re-applying + fixing issues

**Comment by sakluger** [comment]
No updates, we're still working on getting this fixed.

**Comment by muttmuure** [comment]
How is this going @hannojg?

**Comment by hannojg** [comment]
New PR will be ready in a moment:

- https://github.com/Expensify/App/pull/51954

**Comment by sakluger** [comment]
Hey @hannojg, what's the latest? Looks like there is a bit more cleanup needed before we can merge the PR you linked above?

**Comment by hannojg** [comment]
Yes exactly, would be awesome if we could work through that list here:

https://github.com/Expensify/App/pull/51954#issuecomment-2464282215

After those are merged I can work on the next set of cleanup PRs. The final goal is to cleanup the existing code so that i can reuse a stable ordering logic for the new performance PR :)

**Comment by sakluger** [comment]
Latest update here: https://github.com/Expensify/App/pull/51954#issuecomment-2485511954

- There is one cleanup PR left, then a few additional changes needed to fix an ordering bug.

**Comment by sakluger** [comment]
The final cleanup PR has been created (more info: https://github.com/Expensify/App/pull/51954#issuecomment-2511170844)

**Comment by sakluger** [comment]
As of yesterday, cleanups are complete. Now Hanno is working on fixing the Search suffix tree implementation PR. Latest update here: https://github.com/Expensify/App/pull/51954#issuecomment-2539512480

**Comment by sakluger** [comment]
Still testing things in the PR.

**Comment by sakluger** [comment]
@hannojg @kirillzyusko is https://github.com/Expensify/App/issues/46591 done? If so, that's super exciting!

What are the next steps here?
- I see some investigations in [[Performance] Opening search for the first time on a big account is causing the app to hang massively](https://github.com/Expensify/App/issues/46590#top)
- What are the plans for the other two issues: [[Performance] After clicking on a search result it took 4x OpenReport requests to open the final report](https://github.com/Expensify/App/issues/46594#top) and https://github.com/Expensify/App/issues/46593

**Comment by sakluger** [comment]
@hannojg checking in again - what are the next steps for this one?

**Comment by sakluger** [comment]
No updates - I'll ask in Slack later this week or next week for the status of the general investigation.

**Comment by sakluger** [comment]
I created a new thread in the Slack `#quality` room to ask what we should do next here. As a start, I'm going to try to collect a masked Onyx export from this admin (by asking them to go to Settings > Troubleshoot > Download Onyx State).

I'll continue the discussion in that thread and update here.

**Comment by hannojg** [comment]
Next step would actually be to check with the same customer the performance has improved (also specifically the search).
While checking we should record new performance profiles.

(Sorry for letting this go stale, I was OOO for a bit)

**Comment by sakluger** [comment]
@hannojg I spoke with the customer, and they are still experiencing some performance issues related to loading the LHN and searching for chats in the LHN. I asked the user to download their Onyx state after experiencing performance issues, which they sent me and I posted here in Slack: https://expensify.slack.com/archives/C05LX9D6E07/p1740180810363059?thread_ts=1739309454.559869&cid=C05LX9D6E07

**Comment by hannojg** [comment]
Thanks for sharing, will take a look!

**Comment by sakluger** [comment]
Any more updates?

**Comment by hannojg** [comment]
Hey, we still have this on our Todo! @Nodonisko is currently finishing up some other task in expensify and will switch to continue investigating the performance here soon!
@chrispader tagging you here to help with scheduling - thanks!

**Comment by sakluger** [comment]
Hey @hannojg, @Nodonisko, and @chrispader, just checking in to see how things are going with this investigation.

**Comment by sakluger** [comment]
Hey @Nodonisko! Any updates? Were you able to make any progress on this?

**Comment by sakluger** [comment]
Good find @Nodonisko! Does it seem like the fast search tree is recreated many times in error, or was that done intentionally?

Please let us know if you find anything else.

**Comment by Nodonisko** [comment]
Sure thing, I will post updates here.

**Comment by Nodonisko** [comment]
After some more investigation there seems to be also memory leak in component `SearchAutocompleteList` when using `getParticipantsAutocompleteList`. But still investigating what's exact reason for that leak.

**Comment by sakluger** [comment]
Hey @Nodonisko, I'm back from OOO. Any additional updates?

**Comment by Nodonisko** [comment]
@sakluger Hi, I recommend closing this one in favor of [#60889](https://github.com/Expensify/App/issues/60889). I handed all my finding from this issue to @sosek108 and he is continuing with further investigation.

**Comment by sakluger** [comment]
Sounds good to me! Closing in favor of https://github.com/Expensify/App/issues/60889.
